pandas/io/parsers/python_parser.py

```diff
@@ -1167,6 +1165,22 @@ def _remove_skipped_rows(self, new_rows: list[list[Scalar]]) -> list[list[Scalar
             ]
         return new_rows
 
+    def _set_no_thousand_columns(self) -> set[int]:
+        no_thousands_columns: set[int] = set()
+        if self.columns and self.parse_dates:
+            assert self._col_indices is not None
+            no_thousands_columns = self._set_noconvert_dtype_columns(
+                self._col_indices, self.columns
+            )
+        if self.columns and self.dtype:
+            assert self._col_indices is not None
+            for i in self._col_indices:
+                if isinstance(self.dtype, dict) and not is_numeric_dtype(
+                    self.dtype.get(self.columns[i], None)
```

Comment:
@MarcoGorelli the issue you mentioned is due to this line.
Since your test case doesn't specify the type of the 2nd and the 3rd column, here the code will treat them as non-numeric.
I think the behavior is reasonable, how is your idea?

Reply:
if you change the thousand to `,` it will work as expected. probably we should raise an exception when `self.thouand == self.decimal` ?

Reply:
I don't think that the second and third columns should be parsed any differently if the dtype of `a` has been specified

As in,
- `pd.read_csv(io.StringIO(data), sep=';', dtype={'a': str}, thousands='.', engine='python')`
- `pd.read_csv(io.StringIO(data), sep=';', thousands='.', engine='python')`
should parse columns `'b'` and `'c'` in the same way

> probably we should raise an exception when self.thouand == self.decimal ?

sounds reasonable - this could be done in a separate PR if you're interested